src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -469,175 +472,235 @@ public SiloAddress CalculateTargetSilo(GrainId grain, bool excludeThisSiloIfStop
                 throw new InvalidOperationException("Grain directory is stopping");
             }
+
             if (owner.Equals(MyAddress))
             {
-                RegistrationsSingleActLocal.Increment();
-                // if I am the owner, store the new activation locally
-                Tuple<ActivationAddress, int> returnedAddress = DirectoryPartition.AddSingleActivation(address.Grain, address.Activation, address.Silo);
-                return returnedAddress == null ? null : returnedAddress.Item1;
+                // if I am the owner, perform the operation locally
+                return null;
             }
-            else
-            {
-                RegistrationsSingleActRemoteSent.Increment();
-                // otherwise, notify the owner
-                Tuple<ActivationAddress, int> returnedAddress = await GetDirectoryReference(owner).RegisterSingleActivation(address, NUM_RETRIES);
-
-                // Caching optimization: 
-                // cache the result of a successfull RegisterSingleActivation call, only if it is not a duplicate activation.
-                // this way next local lookup will find this ActivationAddress in the cache and we will save a full lookup!
-                if (returnedAddress == null || returnedAddress.Item1 == null) return null;
-
-                if (!address.Equals(returnedAddress.Item1) || !IsValidSilo(address.Silo)) return returnedAddress.Item1;
 
-                var cached = new List<Tuple<SiloAddress, ActivationId>>(1) { Tuple.Create(address.Silo, address.Activation) };
-                // update the cache so next local lookup will find this ActivationAddress in the cache and we will save full lookup.
-                DirectoryCache.AddOrUpdate(address.Grain, cached, returnedAddress.Item2);
-                return returnedAddress.Item1;
+            if (hopCount >= HOP_LIMIT)
+            {
+                // we are not forwarding because there were too many hops already
+                throw new OrleansException(string.Format("Silo {0} is not owner of {1}, cannot forward {2} to owner {3} because hop limit is reached", MyAddress, grainId, operationDescription, owner));
             }
+
+            // forward to the silo that we think is the owner
+            return owner;
         }
 
-        public async Task RegisterAsync(ActivationAddress address)
+
+        public async Task<AddressAndTag> RegisterAsync(ActivationAddress address, bool singleActivation, int hopCount)
         {
-            registrationsIssued.Increment();
-            SiloAddress owner = CalculateTargetSilo(address.Grain);
-            if (owner == null)
+            var counterStatistic = 
+                singleActivation 
+                ? (hopCount > 0 ? this.RegistrationsSingleActRemoteReceived : this.registrationsSingleActIssued)
+                : (hopCount > 0 ? this.RegistrationsRemoteReceived : this.registrationsIssued);
+
+            counterStatistic.Increment();
```

Comment:
It think it is right. To clarify: this single method is a de-duplication of code that used to be in the **four** method bodies

`RemoteGrainDirectory.RegisterSingleActivation`
`RemoteGrainDirectory.Register`
`LocalGrainDirectory.RegisterSingleActivationAsync`
`LocalGrainDirectory.RegisterAsync`
 
it is supposed to do the exact same things as those did before. In order to know which original method we correspond to, we can 
1. use `hopcount` to distinguish between LocalGrainDirectory (hopcount = 0) and RemoteGrainDirectory (hopcount > 0) 
2. use `singleActivation` to distinguish between the two versions of activation

this line does exactly that - it uses the same counters that are in those four original methods